loki/transformations/inline/functions.py

```diff
@@ -0,0 +1,342 @@
+         Procedure in which to inline functions.
+    inline_elementals_only : bool, optional
+        Inline elemental routines/functions only (default: False).
+    functions : tuple, optional
+        Inline only functions that are provided here
+        (default: None, thus inline all functions).
+
+    Returns
+    -------
+    bool
+        Whether inline calls are (potentially) left to be
+        inlined in the next call to this function.
+    """
+
+    class ExpressionRetrieverSkipInlineCallParameters(ExpressionRetriever):
+        """
+        Expression retriever skipping parameters of inline calls.
+        """
+        # pylint: disable=abstract-method
+
+        def __init__(self, query, recurse_query=None, inline_elementals_only=False,
+                functions=None, **kwargs):
+            self.inline_elementals_only = inline_elementals_only
+            self.functions = as_tuple(functions)
+            super().__init__(query, recurse_query, **kwargs)
+
+        def map_inline_call(self, expr, *args, **kwargs):
+            if not self.visit(expr, *args, **kwargs):
+                return
+            self.rec(expr.function, *args, **kwargs)
+            # SKIP parameters/args/kwargs on purpose
+            #  under certain circumstances
+            if expr.procedure_type is BasicType.DEFERRED or\
+                    (self.inline_elementals_only and\
+                    not(expr.procedure_type.is_function and expr.procedure_type.is_elemental)) or\
+                    (self.functions and expr.routine not in self.functions):
+                for child in expr.parameters:
+                    self.rec(child, *args, **kwargs)
+                for child in list(expr.kw_parameters.values()):
+                    self.rec(child, *args, **kwargs)
+
+            self.post_visit(expr, *args, **kwargs)
+
+    class FindInlineCallsSkipInlineCallParameters(ExpressionFinder):
+        """
+        Find inline calls but skip/ignore parameters of inline calls.
+        """
+        retriever = ExpressionRetrieverSkipInlineCallParameters(lambda e: isinstance(e, sym.InlineCall))
+
+    # functions are provided, however functions is empty, thus early exit
+    if functions is not None and not functions:
+        return False
+    functions = as_tuple(functions)
+
+    # Keep track of removed symbols
+    removed_functions = set()
+
+    # Find and filter inline calls and corresponding nodes
+    function_calls = {}
+    # Find inline calls but skip/ignore inline calls being parameters of other inline calls
+    #  to ensure correct ordering of inlining. Those skipped/ignored inline calls will be handled
+    #  in the next call to this function.
+    retriever = ExpressionRetrieverSkipInlineCallParameters(lambda e: isinstance(e, sym.InlineCall),
+            inline_elementals_only=inline_elementals_only, functions=functions)
+    # override retriever ...
+    FindInlineCallsSkipInlineCallParameters.retriever = retriever
```

Comment:
I must have missed this before: This shouldn't be necessary, since the `FindInlineCallsSkipInlineCallParameters` is declared inside the function, we should be able to directly instantiate the retriever with the provided values for `inline_elementals_only` and `functions`.

Reply:
Yes, now I understand. Pushed a fix to this extend.